spec/mailers/application_mailer_spec.rb

```diff
@@ -90,7 +90,7 @@ def create_multipart_method(method_name)
         create_multipart_method('multipart_theme_only')
         @mail = ApplicationMailer.multipart_theme_only
         expect(@mail.parts.size).to eq(2)
-        message_part = @mail.parts[0].to_s
+        message_part = @mail.parts[1].to_s
```

Comment:
There's quite a bit of change here – not just ordering. Not sure whether this is expected or not.

## Develop

Here's a spec that compares the part content against a fixed string.

To get the strings I opened a pry session and called `@mail.parts[N].to_s` and copied the result into the spec.

```diff
diff --git a/spec/mailers/application_mailer_spec.rb b/spec/mailers/application_mailer_spec.rb
index 209147b4b..7ea5f1e71 100644
--- a/spec/mailers/application_mailer_spec.rb
+++ b/spec/mailers/application_mailer_spec.rb
@@ -90,6 +90,12 @@ describe ApplicationMailer do
         create_multipart_method('multipart_theme_only')
         @mail = ApplicationMailer.multipart_theme_only
         expect(@mail.parts.size).to eq(2)
+
+        part_a = "Content-Type: text/html;\r\n charset=UTF-8\r\nContent-Transfer-Encoding: 7bit\r\n\r\nTheme multipart"
+        part_b = "Content-Type: message/rfc822;\r\n filename=original.eml\r\nContent-Transfer-Encoding: base64\r\nContent-Disposition: attachment;\r\n filename=original.eml\r\nContent-ID: <5f297369bf553_35fe2acc93c3a5c0744c7@alaveteli-bionic64.mail>\r\n\r\neHh4\r\n"
+        parts = [@mail.parts[0].to_s, @mail.parts[1].to_s]
+        expect(parts).to eq([part_a, part_b])
+
         message_part = @mail.parts[0].to_s
         expect(message_part).to match("Theme multipart")
       end
```

```
Failures:

  1) ApplicationMailer when using plugins when a plugin prepends its mail templates to the view paths should render a multipart email using a theme template
     Failure/Error: expect(parts).to eq([part_a, part_b])

       expected: ["Content-Type: text/html;\r\n charset=UTF-8\r\nContent-Transfer-Encoding: 7bit\r\n\r\nTheme multipar...l.eml\r\nContent-ID: <5f297369bf553_35fe2acc93c3a5c0744c7@alaveteli-bionic64.mail>\r\n\r\neHh4\r\n"]
            got: ["Content-Type: text/html;\r\n charset=UTF-8\r\nContent-Transfer-Encoding: 7bit\r\n\r\nTheme multipar...l.eml\r\nContent-ID: <5f29752ae3189_36612ac68ac2e5b863959@alaveteli-bionic64.mail>\r\n\r\neHh4\r\n"]

       (compared using ==)

       Diff:
       @@ -1,3 +1,3 @@
        Content-Type: text/html;\r\n charset=UTF-8\r\nContent-Transfer-Encoding: 7bit\r\n\r\nTheme multipart
       -Content-Type: message/rfc822;\r\n filename=original.eml\r\nContent-Transfer-Encoding: base64\r\nContent-Disposition: attachment;\r\n filename=original.eml\r\nContent-ID: <5f297369bf553_35fe2acc93c3a5c0744c7@alaveteli-bionic64.mail>\r\n\r\neHh4\r\n
       +Content-Type: message/rfc822;\r\n filename=original.eml\r\nContent-Transfer-Encoding: base64\r\nContent-Disposition: attachment;\r\n filename=original.eml\r\nContent-ID: <5f29752ae3189_36612ac68ac2e5b863959@alaveteli-bionic64.mail>\r\n\r\neHh4\r\n
```

The only thing that differs is the Content-iD header. That's fine, since in these tests we're generating the mail.

## 2f964e4a2

Same spec as before, except we swap the part order.

```diff
- expect(parts).to eq([part_a, part_b])
+ expect(parts).to eq([part_a, part_b])
```

We're still expecting the same string as in `develop`.

```diff
diff --git a/spec/mailers/application_mailer_spec.rb b/spec/mailers/application_mailer_spec.rb
index 209147b4b..7ea5f1e71 100644
--- a/spec/mailers/application_mailer_spec.rb
+++ b/spec/mailers/application_mailer_spec.rb
@@ -90,6 +90,12 @@ describe ApplicationMailer do
         create_multipart_method('multipart_theme_only')
         @mail = ApplicationMailer.multipart_theme_only
         expect(@mail.parts.size).to eq(2)
+
+        part_a = "Content-Type: text/html;\r\n charset=UTF-8\r\nContent-Transfer-Encoding: 7bit\r\n\r\nTheme multipart"
+        part_b = "Content-Type: message/rfc822;\r\n filename=original.eml\r\nContent-Transfer-Encoding: base64\r\nContent-Disposition: attachment;\r\n filename=original.eml\r\nContent-ID: <5f297369bf553_35fe2acc93c3a5c0744c7@alaveteli-bionic64.mail>\r\n\r\neHh4\r\n"
+        parts = [@mail.parts[0].to_s, @mail.parts[1].to_s]
+        expect(parts).to eq([part_b, part_a])
+
         message_part = @mail.parts[0].to_s
         expect(message_part).to match("Theme multipart")
       end
```

```
Failures:

  1) ApplicationMailer when using plugins when a plugin prepends its mail templates to the view paths should render a multipart email using a theme template
     Failure/Error: expect(parts).to eq([part_b, part_a])

       expected: ["Content-Type: message/rfc822;\r\n filename=original.eml\r\nContent-Transfer-Encoding: base64\r\nCon...ntent-Type: text/html;\r\n charset=UTF-8\r\nContent-Transfer-Encoding: 7bit\r\n\r\nTheme multipart"]
            got: ["Content-Type: message/rfc822;\r\n filename=original.eml\r\nContent-Transfer-Encoding: \r\nContent-D...tent-Type: text/plain;\r\n charset=UTF-8\r\nContent-Transfer-Encoding: 7bit\r\n\r\nTheme multipart"]

       (compared using ==)

       Diff:
       @@ -1,3 +1,3 @@
       -Content-Type: message/rfc822;\r\n filename=original.eml\r\nContent-Transfer-Encoding: base64\r\nContent-Disposition: attachment;\r\n filename=original.eml\r\nContent-ID: <5f297369bf553_35fe2acc93c3a5c0744c7@alaveteli-bionic64.mail>\r\n\r\neHh4\r\n
       -Content-Type: text/html;\r\n charset=UTF-8\r\nContent-Transfer-Encoding: 7bit\r\n\r\nTheme multipart
       +Content-Type: message/rfc822;\r\n filename=original.eml\r\nContent-Transfer-Encoding: \r\nContent-Disposition: attachment;\r\n filename=original.eml\r\nContent-ID: <5f29764bb3ac4_36962ae22aa665bc469b4@alaveteli-bionic64.mail>\r\n\r\nxxx
       +Content-Type: text/plain;\r\n charset=UTF-8\r\nContent-Transfer-Encoding: 7bit\r\n\r\nTheme multipart
```

So what's different:

1. `Content-Transfer-Encoding` in the first part was `base64` and is now empty
2. It looks like the body of the first part was `eHh4` and is now `xxx`
3. `Content-Type` of the second part was `text/html` but is now `text/plain`

Reply:
You're correct there is more going on here.

First, `xxx` encoded as base64 is `eHh4` so should be expect given the `Content-Transfer-Encoding` now isn't `base64` but is that correct...

I've saved the spec example mails to see if they appear correct in Mail.app...

# Previous

![image](https://user-images.githubusercontent.com/5426/89513973-c6608e80-d7c4-11ea-8db8-9c1d6efde3d2.png)
And the attachment is correct:
```
$ cat original.eml
xxx
```

# New

![image](https://user-images.githubusercontent.com/5426/89514382-4dae0200-d7c5-11ea-9115-752b1154bb23.png)
And no attachment detected.

Will investigate further.

Reply:
I have been testing this issue using. The same problem occurs with a new/bare Rails 6 or Rails edge application.

```ruby
class TestMailer < ActionMailer::Base
  def test
    attachments['message.eml'] = 'xxx'
    mail { |format| format.text { render plain: 'Test' } }
  end
end
File.open('tmp/test.eml', 'w') { |f| f.write TestMailer.test }
```

Reply:
To bypassing `ActionMailer` we can do:

```ruby
mail = Mail.new do
  subject 'Test'
  text_part do
    body 'Test'
  end
  attachments['message.eml'] = 'xxx'
end
File.open('tmp/test-2.eml', 'w') { |f| f.write mail }
```

This mail is okay

Reply:
With message ids and mail part boundaries normalised, the resulting diff is:

```diff
--- ./tmp/test-1.eml	2020-08-06 13:24:09.000000000 +0100
+++ ./tmp/test-2.eml	2020-08-06 13:25:37.000000000 +0100
@@ -2,7 +2,7 @@
 Message-ID: <5f2bf669bc448_126853ffc1683404436914@localhost>
 Subject: Test
 Mime-Version: 1.0
-Content-Type: text/plain;
+Content-Type: multipart/mixed;
  boundary="--==_mimepart_5f2bf669bbb50_126853ffc1683404436842";
  charset=UTF-8
 Content-Transfer-Encoding: 7bit
```

Reply:
Looking the Rails source `multipart/mixed` [should be set](https://github.com/rails/rails/blob/d08f2d4/actionmailer/lib/action_mailer/base.rb#L877) but `m.has_attachments?` is `false` This is calling a [method in the mail gem](https://github.com/mikel/mail/blob/a80a251/lib/mail/message.rb#L1669), so it would seem there is something going on relating to attaching `.eml`.

I think, as its an email, it is trying to [pull out the internal parts of the attached file](https://github.com/mikel/mail/blob/a80a251/lib/mail/attachments_list.rb#L10) - but in our case its not a real email so it doesn't work.

Expanding our ActiveMailer example above to:
```ruby
class TestMailer < ActionMailer::Base
  def test
    message = ::Mail.new do
      subject 'Inner'
      text_part do
        body 'Inner'
      end
    end
    attachments['message.eml'] = message.to_s
    mail { |format| format.text { render plain: 'Test' } }
  end
end
File.open('tmp/test-1.eml', 'w') { |f| f.write TestMailer.test }
```
and we get a `multipart/mixed` mail which appears correct in Mail.app.

I'll push a fixup to force `multipart/mixed` in our specs.

Reply:
Fixed in b6d8f58

Reply:
Looks good! Nice work!

Tested a few variations of attaching real emails to emails and ingesting them and all works fine between `develop` and this branch.

![Screenshot 2020-08-11 at 10 24 47](https://user-images.githubusercontent.com/282788/89881129-f9d45c00-dbbc-11ea-8c13-27ebbf5dcd4a.png)

![Screenshot 2020-08-11 at 10 25 50](https://user-images.githubusercontent.com/282788/89881180-0d7fc280-dbbd-11ea-9f7e-9b921324681d.png)

![Screenshot 2020-08-11 at 10 26 07](https://user-images.githubusercontent.com/282788/89881207-17a1c100-dbbd-11ea-96fd-6a2c9f7d4512.png)

Won't bore you with them all, but I'm pretty confident that its doing what we were expecting.